streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1653,7 +1654,9 @@ public boolean equals(final Object o) {
 
             final Subtopology that = (Subtopology) o;
             return id == that.id
-                && nodes.equals(that.nodes);
+                // convert both TreeSets to arrays to ensure .equals() is used recursively
+                // otherwise, the provides NODE_COMPARATOR is used what might result in incorrect comparison
+                && Arrays.equals(nodes.toArray(), that.nodes.toArray());
```

Comment:
We actually have sufficient test coverage for `topicPattern==null`, however, because the original code here is executed in an unexpected way, `Source.equals()` is not called: Because we pass in `NodeComperator` into the `TreeSet`, the comparator is use in `TreeMap#equals()`. However, the comparator uses a "shortcut" instead of doing a deep comparison. This "shortcut" is fine for `InternalTopologyBuilder` but breaks a correct "deep comparison" of `TopologyDescription`. Switching to array comparison skips the `NodeComperator` and provides correct "deep comparison" via `Node#equals()`.

With this change (plus change below) and without the fix from above, we hit the NPE.